text/0000-typed-context-injection.md

```diff
@@ -0,0 +1,833 @@
+
+```rust
+fn func_1(..., new_system: &mut NewSystem) {  // Updated!
+    ...
+    func_2(..., new_system);  // Updated!
+}
+
+fn func_2(..., new_system: &mut NewSystem) { // Updated!
+    ...
+    func_3(..., new_system);  // Updated!
+}
+
+fn func_3(..., new_system: &mut NewSystem) { // Updated!
+    // We finally have access to `new_system`
+}
+```
+
+Although one may be tempted to bundle up this context into convenient tuples to reduce the amount of explicit context passing, doing so will only work if at most one descendant of the call stack relies upon that type. If several descendants rely on a given contextual type, you're out of luck!
+
+```rust
+type Descendant1Cx<'a> = (&'a mut SharedSystem, &'a mut Descendant1System);
+
+fn descendant_1(cx: Descendant1Cx<'_>) {
+    ...
+}
+
+type Descendant2Cx<'a> = (&'a mut SharedSystem, &'a mut Descendant2System);
+
+fn descendant_2(cx: Descendant2Cx<'_>) {
+    ...
+}
+
+fn callee(descendant_1_cx: Descendant1Cx<'_>, descendant_2_cx: Descendant2Cx<'_>) {
+    descendant_1(descendant_1_cx);
+    descendant_2(descendant_2_cx);
+}
+
+fn caller(
+    shared_system: &mut SharedSystem,
+    descendant_1_system: &mut Descendant1System,
+    descendant_2_system: &mut Descendant2System,
+) {
+    callee(
+        (shared_system, descendant_1_system),
+        (shared_system, descendant_2_system),  // Whoops! We borrowed shared_system twice!
+    );
+}
+```
+
+Finally, although good software practices encourage programmers to keep their abstractions minimal and properly-decoupled, doing so only causes grief when trying to implement operations acting upon several of these well-split-up components:
+
+```rust
+pub struct AudioBuffer { ... }
+
+impl AudioBuffer {
+    pub fn play_samples(&mut self, data: &[f32], volume: f32) { ... }
+}
+
+pub struct VirtualFileSystem { ... }
+
+impl VirtualFileSystem {
+    pub fn load_file(&mut self, path: &str) -> &[u8] { ... }
+}
+
+pub struct AudioCache { ... }
+
+impl AudioCache {
+    pub fn resolve_from_cache(&mut self, path: &str, loader: impl FnOnce() -> &[u8]) -> &[f32] { ... }
+}
+
+pub struct SoundVolumeOverrides { ... }
+
+impl SoundVolumeOverrides {
+    pub fn get_volume_override(path: &str) -> f32 { ... }
+}
+
+// Ideally, we could just write:
+cx.play_sound_at_path("cbat.ogg");
+
+// But, unfortunately, we must write:
+audio_buffer.play_sound_at_path(&mut virtual_fs, &mut audio_cache, &mut sound_volume_overrides, "cbat.ogg");
+```
+
+Yuck!
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Enter typed context injection. 
+
+Context injection is accomplished with two new standard library items: `Cx` and `AnyCx`. `Cx` is a variadic type parameterized by the references comprising a function's context. For example, to write a function taking `&mut System1` and `&System2`, you can just write:
```

Comment:
Yes, it is indeed possible to make `Cx` take in a single tuple and provide some nice syntactic sugar to transform `Cx(&Type1, &Type2)` into `Cx<(&Type1, &Type2)>`. I wrote things this way because I thought it would be easier to implement that way but I am open to changing it if other Rust developers agree that doing so would be better.